src/google/protobuf/compiler/java/java_message.cc

```diff
@@ -1446,13 +1446,17 @@ void ImmutableMessageGenerator::GenerateAnyMethods(io::Printer* printer) {
     "public <T extends com.google.protobuf.Message> T unpack(\n"
     "    java.lang.Class<T> clazz)\n"
     "    throws com.google.protobuf.InvalidProtocolBufferException {\n"
-    "  if (!is(clazz)) {\n"
+    "  boolean invalidClazz = false;\n"
```

Comment:
What if previously an instance of the same class but different value was unpacked?

Reply:
Old implementation would still return `cachedUnpackValue` in this hypothetical case, this implementation retains this behavior.